src/lang/en/man/proxytester.1
.\" Proxytester man page created by Sam Gleske Mon Mar 29 2010
.TH proxytester 1 "March 29, 2010" PROXYTESTER "Proxytester\-0.6"
.\"
.\" NAME part
.\"
.SH NAME
proxytester \- a network utility written in Python to test proxy servers from a proxy list and determines if they are working
.\"
.\" SYNOPSIS part
.\"
.SH SYNOPSIS
.BI "proxytester [\-O " outFile "] " fileList
.br
.BI "proxytester [\-GET " url "] [\-E] [\-MT] [\-Q] [\-T " timeout "] [\-U] [\-W]
.br
.BI "            [\-O " outFile "] [\-RESPONSE " inFile "] " fileList
.br
.B proxytester \-\-help
.br
.B proxytester \-\-license
.\"
.\" DESCRIPTION part
.\"
.SH DESCRIPTION
.B Proxytester
takes a proxy list as input and then tests the addresses to ensure that they are still available for use. This program can also generate a wpad.dat file which can be used by browsers. The wpad.dat file uses the working list of proxies and each time the browser makes a request it will select a random proxy which makes the user virtually impossible to track.
.\"
.\" PARAMETER part
.\"
.SH PARAMETERS
.TP
.I fileList
1 or more file paths to a proxy list.
.TP
.BR /?
Shows the help dialog.  Also \-help and \-\-help work.
.TP
.BR \-GET
Specify a link to test the proxy server against.
.br
If \-GET url is not specified then this will get Google.com.
.TP
.BR \-E
Proxy list of proxy servers to exclude from testing. They will be excluded if they're in this list even if they work.  
.TP
.IR inFile
File path to plain text file to be read.
.TP
.BR \-license
Display the MIT License. Also \-\-license works.
.TP
.BR \-MT
Multi\-Threading support to process proxy lists faster.       
.TP
.BR \-O
Outputs a proxy list of working proxies based on given options.
.TP
.IR outFile
File path where output is written to a plain text file.      
.TP
.BR \-Q
Suppress all overwrite prompts.                              
.TP
.BR \-RESPONSE
Read and compare the server response with the input file to ensure the proxy server is working and not just showing any HTML page.
.TP
.BR \-T
Manually set the HTTP timeout for proxy servers.             
.TP
.IR timeout
Timeout number in seconds. The shorter the timeout the faster the proxies must be to respond.                              
.br
    1 \- or lower for high performance proxies
.br
    3 \- average performance proxies
.br
    5 \- or more for slower proxies
.br
    Default is 3. Any range in between numbers work.
.TP
.BR \-U
Creates a unique list of proxy servers with no duplicates.
.TP
.IR url
Internet URL to test for retrieval by the proxy server. This is usually a restricted URL that is being blocked.
.br
If \-GET url is not specified then this will get Google.com.
.TP
.BR \-W
Generates a wpad.dat file.                                   
.\"
.\" Proxy List Part
.\"
.SH PROXY LIST
A proxy list is a plain text file. In the proxy list each new line must contain a single server with the following format.                           
.br
.IR server:port
.\"
.\" EXAMPLES part
.\"
.SH EXAMPLES
.TP
Test proxy servers and generate a new list (Minimum arguments)
.br
.B proxytester \-o
.I new_proxy_list.txt proxylist1.txt
.TP
Test proxy servers, generate a new list, and excluding testing some servers. 
.br
.BI "proxytester " proxylist1.txt " \-e " exclude.txt " \-o " new_proxy_list.txt
.TP
Test proxy servers against a response to ensure that they really work.
.br
.BI "proxytester \-get " http://www.example.com/proxytest.html " \-response " ./proxytest.html " \-o " new_proxy_list.txt " "  proxylist1.txt
.PP
The \-RESPONSE switch is normally used with the \-GET switch but not always.  Sometimes a user wishes to test if the proxy is truly a transparent proxy so the \-RESPONSE switch was devised to ensure this.  Just create your own dummy html page and upload it to your own web server.  Then call that page using the \-GET switch and check it against a local copy of the same html page using the \-RESPONSE switch. This is the best method for testing anonymous proxies\!
.PP
Options don't have to be in any particular order as long as the command format 
is followed. Refer to the Command Format to see what switches are followed by  
arguments.
.\"
.\" AUTHOR PART
.\"
.SH AUTHOR
Sam Gleske \- sam.mxracer@gmail.com
.PP
Copyright (c) 2009 Proxy Tester Project, Sam Gleske
.br
Licensed under the MIT License (see command line options for viewing)
.br
Project website: 
.I http://sourceforge.net/projects/proxytest
.\"
.\" FILES part
.\"
.SH FILES





changelog.txt
This software was written in Python and JavaScript!

Need help?  Check out the wiki...
http://sourceforge.net/apps/mediawiki/proxytest/

Still need help?  Ask in the forums...
http://sourceforge.net/projects/proxytest/forums/

v0.7.1 May 24, 2010
1.	Fixed bug in wpad.dat which was introduced in v0.7
	accidentally where wpad.dat threw an exception only creating
	direct connections and never using the proxies.
2.	Fixed bug in -sim switch where null lines were tested in the
	simulation.  This additionally threw errors with wpad.dat.

v0.7 May 18, 2010
1.	New Feature: proxytester command no longer dependent on -o 
	switch.
2.	New Feature: -sim switch to simulate working proxies without
	connecting to the internet.  This makes it easy to generate
	wpad.dat files from existing proxy lists quickly.  This 
	feature also makes automatic generating of an excludes list
	very easy where this was not possible in previous versions.
3.	Made a few efficiency changes to the -response switch on how
	it strips carriage returns.
4.	Updated the command line documentation adding the title of the
	software and the version number to it.  Also updated adding
	-sim switch to documentation with command examples.
5.	Fixed bug in wpad.dat script where it was possible for a null 
	proxy to be selected.
6.	Fixed bug in UniqueList.py where __init__() function was 
	returning a value.

v0.6 Dec 15, 2009
1.	New feature: excludes list option [-E inFile]
	Excludes list is a proxy list which proxytester excludes from
	testing and assumes to be bad.
2.	New feature: custom response [-RESPONSE inFile]
	Custom response is a remote page to be read and compared to. If
	the page is different then the proxy server is considered bad 
	but if the page is the same then the proxy server is good. I got
	the idea to do this because I kept getting proxy servers that 
	required some kind of credentials in an HTML format but if you
	test against a page response then it means the proxy server is
	more likely to be working. [-RESPONSE inFile] is normally used
	with [-GET url] but not always.
3.	Updated the command line documentation to reflect additional 
	features as well as improved the documentation with examples
	and more descriptions.

v0.5 Dec 6, 2009
1.	Now there is no internal configurations.  Only takes arguments.
2.	Windows binary executable now available.
3.	Can add more than 1 file as input at the end of the line instead
	of just processing a single file.
4.	New features: Optional multithreading, generate wpad.dat, set 
	proxy timeout, custom URL to test proxy servers against.  The
	custom URL is for purposes such as attempting to access a
	restricted site.
5.	Fixed bug where pressing Ctrl+C did not stop the program.


v0.4 Dec 03, 2009
1.	Now uses multithreading to process requests faster.  Processes
	4 requests at a time like a web browser.  Now it processes 
	2000 proxy servers in around 23 minutes when before it took
	hours.
2.	Added a new library for pooling thread requests.
3.	Now with timestamping.  Shows you when the script started, when 
	it ended, and how long it ran.
4.	Other minor bug fixes
5.	Updated the rules for the generated wpad.dat
6.	Updated generate_wpad.bat to remove binary pyc files


v0.3 Dec 01, 2009
1.	MAJOR bug fixes for excluding protocols in wpad-head.js. Excludes 
	weren't previously working at all!
2.	Simplified some methods by defining additional functions in 
	wpad-head.js
3.	Further optimized the process of request testing for excludes by 
	reordering when the excludes are called so proxying is now faster!
4.	Implemented advanced logging so if a user encounters errors they 
	can report it to me with sufficient information.
5.	Added a few configuration options to wpad.dat so users can easily
	download and customize it for themselves own.
6.	Added thorough documentation to wpad.dat so users can easily 
	understand the options
7.	Created some bookmarklets to assist mining a list of proxies from 
	online sources which are located in the lists folder.

v0.2 Dec 01, 2009
1.	Added a new option RESTRICTED_URL to check-status-proxy-address.py
2.	Spacing change to wpad-foot.js

v0.1 Nov 30, 2009
1.	First Release

src/proxytester.py
#!/usr/bin/env python
'''
    Created Dec 1, 2009
    Main driver for application
    Author: Sam Gleske
'''


import urllib2
import socket
import sys
import os.path
from lib.ThreadPool import *
from lib.SwitchParser import *
from lib.UniqueList import UniqueList
from lib.GenerateWPAD import *
from time import sleep
from time import ctime
from sys import exit

TAB = "\t"
NEW_LINE = "\n"
wpad=GenerateWPAD()

""" make configuration adjustments based on given arguments """
config=SwitchParser(sys.argv)

print "Generate wpad.dat:", str(config.WPAD)
if config.Threads > 1:
    print "Multi-Threading: Enabled"
else:
    print "Multi-Threading: Disabled"
if config.simulateConnect :
    print "Simulate Working Proxies: On"
else:
    print "Simulate Working Proxies: Off"
if config.excludeFile == None :
    print "Exclude certain proxies: False"
else:
    print "Exclude certain proxies: False"
print "Testing URL:", config.restrictedURL
print "Check proxy response against:",config.responseFile
print "Proxy Timeout:", str(config.Timeout)
print "Unique List:", str(config.unique)

#remove duplicate file entries

if len(config.fileList) != 0 :
    config.fileList=UniqueList(config.fileList).unique
else:
    print "Must specify at least one proxy list file."
    config.syntaxErr()


#test to make sure all files exist
for filename in config.fileList:
    if not os.path.isfile(filename):
        print "All files in your fileList must exist!"
        config.syntaxErr()
    elif filename == config.outFile:
        print "One of your fileList files are the same as your outFile"
        config.syntaxErr()
if not config.quietMode :
    if config.outFile != None :
        if os.path.isfile(config.outFile) :
            answer=raw_input("It appears your outFile already exists!" + NEW_LINE + "Do you want to overwrite (Y/N)?: ")
            if answer.upper() not in ('Y','YE','YES') :
                print "User aborted command!"
                exit()

#testing swich accuracy only
#print "config.outFile: " + config.outFile
#print "config.fileList: " + str(config.fileList)
#print "config.WPAD: " + str(config.WPAD)
#print "config.Threads: " + str(config.Threads)
#print "config.quietMode: " + str(config.quietMode)
#print "config.restrictedURL: " + config.restrictedURL
#print "config.Timeout: " + str(config.Timeout)
#print "config.unique:", str(config.unique)
#exit()

# the status will be True if the proxy is good for use!
def checkProxy(pip):
    status=-1
    if config.simulateConnect :
        if pip != "" :
            print pip, "is working"
            status = True
        else :
            status = False
        return (pip, status)
    try:
        proxy_handler = urllib2.ProxyHandler({'http': pip})
        opener = urllib2.build_opener(proxy_handler)
        opener.addheaders = [('User-agent', 'Mozilla/5.0')]
        urllib2.install_opener(opener)
        req = urllib2.Request(config.restrictedURL)
        sock = urllib2.urlopen(req)
        if config.Response != None:
            response = sock.read();
            response = response.strip('\r')
            responseList = []
            responseList = response.split(NEW_LINE)
            for line in responseList :
                if line not in config.Response:
                    status = False
        if not status :
            print "ERROR: Response test failed"
            print "Bad Proxy", pip
    except urllib2.HTTPError, e:
        print 'Error code: ', e.code
        print "Bad Proxy", pip
        status = False
    except Exception, detail:
        print "ERROR:", detail
        print "Bad Proxy", pip
        status = False
    finally:
        if status == -1:
            print pip, "is working"
            status = True
    return (pip, status)

#handles the checkProxy class for threading
def handleResult(result):
#    print result[0], result[1]
    if result[1]:
        tested_proxies.append(result[0])

def waitTask(timeout):
#    print "timeout executed"
    sleep(timeout)


started = ctime()
socket.setdefaulttimeout(config.Timeout)
pool = ThreadPool(config.Threads)

# read the list of proxy IPs in proxyList
proxyList=[]
for filepath in config.fileList:
        f=open(filepath, 'r')
        fileContents = f.read()
        contentsList = fileContents.split(NEW_LINE)
        f.close()
        for line in contentsList:
            proxyList.append(line)
if config.unique :
    proxyList=UniqueList(proxyList).unique

if config.WPAD:
    #test for wpad overwrite
    if not config.quietMode :
        if os.path.isfile('wpad.dat') :
            answer=raw_input("It appears your wpad.dat file already exists!" + NEW_LINE + "Do you want to overwrite (Y/N)?: ")
            if answer.upper() not in ('Y','YE','YES') :
                print "User aborted command!"
                exit()

    f = open('wpad.dat', 'w')

    #write the wpad header
    for line in wpad.head:
        f.write(line)

if config.outFile != None :
    n = open(config.outFile, 'w')

# test the proxy list and generate a new list consisting of working proxies
firstline = False
tested_proxies = []

#Create a new thread for each proxy to be checked
for line in proxyList:
    if line not in config.excludeServers :
        pool.queueTask(checkProxy, line, handleResult)

#This while loop is to account for an unknown bug in ThreadPool.py
i=0
while i < config.Threads:
    pool.queueTask(waitTask, config.Timeout + 2)
    i = i+1

pool.joinAll()

for item in tested_proxies:
    item=str(item)
    if config.outFile != None :
        n.write(item + NEW_LINE)
    if config.WPAD:
        if not firstline:
            f.write('"' + item + '"')
            firstline = True
        else:
            f.write(',' + NEW_LINE + TAB + TAB + TAB + '"' + item + '"')

#write the wpad footer
if config.WPAD:
    for line in wpad.foot:
        f.write(line)
if config.outFile != None :
    n.close()
if config.WPAD:
    f.close()
ended = ctime()
print "Process Started:", started
print "Process Ended:", ended
tsplit=str(ended).split(" ")[3].split(":")
ended=int(tsplit[0]) * 3600 + int(tsplit[1]) * 60 + int(tsplit[2])
tsplit=str(started).split(" ")[3].split(":")
started=int(tsplit[0]) * 3600 + int(tsplit[1]) * 60 + int(tsplit[2])
secs=ended-started
#Print the runtime in # hrs # mins # secs
print "Runtime:",secs/3600,"hrs",secs/60 - secs/3600*60,"mins",60*(secs%60)/100,"secs",NEW_LINE
print "Please wait..."
# if user force exits the program then force kill all threads
pool.joinAll(False,False)
exit(0)
